internal/fs/dir_handle.go

```diff
@@ -242,42 +248,53 @@ func (dh *dirHandle) ensureEntries(ctx context.Context) (err error) {
 //
 // LOCKS_REQUIRED(dh.Mu)
 // LOCKS_EXCLUDED(du.in)
-func (dh *dirHandle) ReadDir(
+func (dh *DirHandle) ReadDir(
 	ctx context.Context,
 	op *fuseops.ReadDirOp) (err error) {
 	// If the request is for offset zero, we assume that either this is the first
 	// call or rewinddir has been called. Reset state.
+
 	if op.Offset == 0 {
-		dh.entries = nil
-		dh.entriesValid = false
-	}
+		dh.Entries = nil
+		dh.EntriesValid = false
+		rec.Lock()
+		rec.length = 0
+		rec.err = nil
+		rec.entryForSorting = fuseutil.Dirent{}
+		rec.Unlock()
+		go dh.FetchEntriesAsync(fuseops.RootInodeID, true)
```

Comment:
how are you calling fetchEntriesAsync after the first gcs call is made.

Reply:
When the first gcs call is made the offset is zero.So the goroutine to fetch data from GCS starts.This fetchEntriesAsync actually runs in a loop and stops only if the continuation token is empty (not the firstcall) .In every run of the loop, when 5k objects are fetched.After the fetched entries are populated into directory handles' Dirent , the routines sends a broadcast to signal the waiting routine that a set of entries have been fetched .Then the main goroutine will read from this dirent to serve kernel requests.